apollo-federation/src/sources/connect/json_selection/js_literal.rs

```diff
@@ -0,0 +1,185 @@
+// A JSLiteral is similar to a JSON value (or serde_json::Value), with the
+// addition of PathSelection as a possible leaf value, so literal expressions
+// passed to -> methods (via MethodArgs) can capture both field and argument
+// values and sub-paths, in addition to constant JSON structures and values.
+
+use apollo_compiler::collections::IndexMap;
+use nom::branch::alt;
+use nom::bytes::complete::tag;
+use nom::character::complete::char;
+use nom::character::complete::one_of;
+use nom::combinator::map;
+use nom::combinator::opt;
+use nom::combinator::recognize;
+use nom::multi::many0;
+use nom::sequence::delimited;
+use nom::sequence::pair;
+use nom::sequence::preceded;
+use nom::sequence::tuple;
+use nom::IResult;
+use serde::ser::SerializeMap;
+use serde::ser::SerializeSeq;
+use serde::Serialize;
+use serde_json_bytes::Value as JSON;
+
+use super::helpers::spaces_or_comments;
+use super::parser::parse_string_literal;
+use super::parser::Key;
+use super::parser::PathSelection;
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum JSLiteral {
+    String(String),
+    Number(String),
+    Bool(bool),
+    Null,
+    Object(IndexMap<String, JSLiteral>),
+    Array(Vec<JSLiteral>),
+    Path(PathSelection),
+}
+
+impl JSLiteral {
+    // JSLiteral   ::= JSPrimitive | JSObject | JSArray | PathSelection
+    // JSPrimitive ::= StringLiteral | JSNumber | "true" | "false" | "null"
+    pub fn parse(input: &str) -> IResult<&str, Self> {
+        tuple((
+            spaces_or_comments,
+            alt((
+                map(parse_string_literal, Self::String),
+                Self::parse_number,
+                map(tag("true"), |_| Self::Bool(true)),
+                map(tag("false"), |_| Self::Bool(false)),
+                map(tag("null"), |_| Self::Null),
+                Self::parse_object,
+                Self::parse_array,
+                // TODO Disallow KeyPath PathSelection in JSLiteral?
+                map(PathSelection::parse, Self::Path),
+            )),
+            spaces_or_comments,
+        ))(input)
+        .map(|(input, (_, value, _))| (input, value))
+    }
+
+    // JSNumber    ::= "-"? (UnsignedInt ("." [0-9]*)? | "." [0-9]+)
+    // UnsignedInt ::= "0" | [1-9] NO_SPACE [0-9]*
+    fn parse_number(input: &str) -> IResult<&str, Self> {
+        delimited(
+            spaces_or_comments,
+            tuple((
+                opt(char('-')),
+                spaces_or_comments,
+                tuple((
+                    recognize(alt((
```

Comment:
This code has changed a bit (for the simpler), but my understanding is that the whole `alt(recognize(...), recognize(...), ...)` combinator produces a single `&str` since each of the `recognize` alternatives is a `&str`.